client_code/routing.py

```diff
@@ -119,31 +125,57 @@ def on_navigation(self, url_hash=None, url_pattern=None, url_dict=None, path=Non
         raise  # this was an unexpected error so raise it
       else:
         self.content_panel.clear()  # clear the form now just incase we end up with a new to cache form that is slow to load later
-        self.load_form(url_hash=url_hash, url_pattern=url_pattern, url_dict=url_dict, path=path)
+        self.load_form(url_hash=url_hash, url_pattern=url_pattern, url_dict=url_dict, path=path, dynamic_vars=dynamic_vars)
       
       _on_navigation_stack_depth -= 1
 
 
     def find_path(self, url_hash, url_pattern, url_dict):
       # this method is called whenever we have form to load from navigation that is not in the cache 
+      given_url_parts = url_pattern.split('/') # determine the individual portions
+      # of this deep link. N.B. this disallows using a '/' within the dynamic variable
+      num_given_url_parts = len(given_url_parts)
+      logger.print(f'[find_path] searching {_paths}')
       for path in _paths:
-        if path.url_pattern == url_pattern and set(url_dict) == set(path.url_keys):
-          return path
+        dynamic_vars = {}
+        target_url_parts = path.url_pattern.split('/')
+        logger.print(f'[find_path] trying {path.url_pattern} against {url_pattern}'
+                     f'; {given_url_parts} -> {target_url_parts}')
+        if num_given_url_parts != len(target_url_parts):
+          # url pattern CANNOT fit, skip deformatting
+          logger.print('[find_path] pattern cannot fit')
+          continue
+        for given, expected in zip(given_url_parts, target_url_parts):
+          logger.print(f'[find_path] testing {given} against {expected}')
+          if expected.startswith('{') and expected.endswith('}'):
+            # dynamic variable
+            logger.print(f'[find_path] found dynamic variable {expected[1:-1]}')
+            dynamic_vars[expected[1:-1]] = given
+          elif given != expected:
+            logger.print(f"[find_path] {expected} isn't a dynamic variable and doesn't match {given}")
+            break
+        else:
+          # given path matches expected path
+          logger.print(f'[find_path] {url_pattern} matches {path.url_pattern}')
+          if set(url_dict) == set(path.url_keys):
+            # only return if the url_dict matches
+            logger.print(f'[find_path] {url_dict} matches {path.url_keys}')
+            return path, dynamic_vars
       else:
```

Comment:
Maybe we can remove some of the logging statements here. I think for most users trying to debug this will be too much information.
Another option would be to create a new logger object like `debug_find_path` and use `debug_find_path.print` so as not to clutter the logging statements.